File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
  const MAX = 102;
  const grid = Array.from({ length: MAX }, () => Array(MAX).fill(-1));
  const visited = Array.from({ length: MAX }, () => Array(MAX).fill(0));

  // 1. 직사각형 테두리 그리기
  for (const rect of rectangle) {
    const [lx, ly, rx, ry] = rect.map(v => v * 2);

    for (let i = lx; i <= rx; i++) {
      for (let j = ly; j <= ry; j++) {
        // 내부 채우기
        if (lx < i && i < rx && ly < j && j < ry) {
          grid[i][j] = 0; // 내부
        } else if (grid[i][j] !== 0) {
          grid[i][j] = 1; // 테두리
        }
      }
    }
  }

  // 2. BFS
  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  const queue = [];
  const startX = characterX * 2;
  const startY = characterY * 2;
  const endX = itemX * 2;
  const endY = itemY * 2;

  queue.push([startX, startY]);

  while (queue.length > 0) {
    const [x, y] = queue.shift();

    if (x === endX && y === endY) {
      return visited[x][y] / 2;
    }

    for (let d = 0; d < 4; d++) {
      const nx = x + dx[d];
      const ny = y + dy[d];

      if (nx > 0 && ny > 0 && nx < MAX && ny < MAX) {
        if (visited[nx][ny] === 0 && grid[nx][ny] === 1) {
          visited[nx][ny] = visited[x][y] + 1;
          queue.push([nx, ny]);
        }
      }
    }
  }

  return 0;
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque
def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 0
    MAX = 102
    grid = [[-1] * MAX for _ in range(MAX)]
    visited = [[0] * MAX for _ in range(MAX)]
    
    for elem in rectangle: 
        lx,ly,rx,ry = map(lambda x:x*2, elem)
        for i in range(lx, rx + 1):
            for j in range(ly, ry + 1):
                if lx < i < rx and ly < j < ry:
                    grid[i][j] = 0
                elif grid[i][j] != 0: # 이미 내부임 -> 테두리 처리 할 필요없음
                    grid[i][j] = 1
                    
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    
    q = deque()
    q.append((characterX * 2, characterY * 2))
    while q:
        x,y = q.popleft()
        if x == itemX * 2 and y == itemY * 2: # 도착
            answer = visited[x][y] // 2
            return answer
        
        for d in range(4):
            nx,ny = x + dx[d], y + dy[d]
            if 0 < nx < MAX and 0 < ny < MAX and not visited[nx][ny]:
                if grid[nx][ny] == 1:
                    visited[nx][ny] = visited[x][y] + 1
                    q.append((nx,ny))
                    
    return answer
    